roachprod, roachtest: add resize operation

herkolategan opened the pull request:
This change adds a new resize operation that will grow clusters and start
cockroach on the new nodes. It copies the current binary and libraries from the
first node, and uses the provided operation config to supply start options and
cluster settings when starting cockroach on the new nodes.

In the new state the operation will pause for a duration before draining and
then decommissioning the new nodes.

If the operation succeeded the cleanup function will shrink the cluster back to
its original size. But if the operation fails, the cleanup function is not
invoked and the cluster will remain at its new size. This is useful for
investigating the new nodes for issues, but also should be monitored since if
the operation is run again it will keep increasing the cluster in size.

Release Note: None
Epic: None

herkolategan committed:
roachprod: add secure flag to grow

Previously, the `grow` command did not take into consideration that if the
cluster is secure the certs would need to be updated with the IP addresses of
the new nodes.

This change takes the `secure` flag into consideration and will update the
certificates and redistribute it to all nodes with the latest information.

Release Note: None
Epic: None

herkolategan committed:
roachprod: fix provider check for grow and shrink

Previously, the provider check could fail if the cloud provider included a
project name. The provider specified on the VM struct itself is what should be
used in the comparison.

Release Note: None
Epic: None

herkolategan committed:
roachtest: add drain and decommission util for operations

Make draining and decommissioning a node available as utility functions to all
operations.

Release Note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126001)
<!-- Reviewable:end -->

renatolabs commented on pkg/cmd/roachtest/operations/resize.go:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	// Start the new nodes.
+	c.Run(ctx, option.WithNodes(newNodes), "./cockroach.sh")
```
For my own understanding, why is it necessary to go through the trouble of editing the startup script (and coupling ourselves to the current implementation) -- what breaks if we use `c.Start` on the new nodes?

herkolategan replied:
We're running into our "roachprod keeps no state" limitation here. In the case of DRT some clusters have additional args passed with the `start` command. We have no way of knowing what those were. One example is the store count, which gets passed. If we were to just invoke start on the new nodes it would not match the rest of the cluster.

**An alternate here may be that we need to supply some yaml / config with an operation runner** so it has some insight about what it is actually running operations against. I see a few new operation on the horizon that look for "tpcc" or make some assumption and then bail out if it is not true. @itsbilal Any thoughts on this?

herkolategan replied:
One update I could make here is to use `Start with Restart`, which essentially just runs the script as well, but we'd still need to either copy & mutate a script from an existing node in the cluster, or store the initial start flags somewhere (could either be passed as flags to run operations, or a config file) - or we could delegate to `drt-run` to pass the flags from a config?

herkolategan replied:
TLDR; I think I'm rather going to expose `StartOpts` as a flag for running operations. In the future the runner of the operation will have to know the details of how the cluster was started.

renatolabs commented on pkg/cmd/roachtest/operations/resize.go:
```diff
@@ -0,0 +1,142 @@
+	time.Sleep(pauseDuration)
+	for i := 0; i < growCount; i++ {
+		drainNode(ctx, o, c, c.Node(1), c.Node(origClusterSize+i+1))
+		decommissionNode(ctx, o, c, c.Node(1), c.Node(origClusterSize+i+1))
```
Should we randomize the "exec" node?

herkolategan replied:
Good idea, I'll add random exec node selection here.

renatolabs reviewed (commented):
Nice!

The `Localities` function added in this PR doesn't seem to be used here?

herkolategan commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2104,6 +2104,26 @@ func (c *clusterImpl) configureClusterSettingOptions(
+
+// Shrink removes nodes from the cluster.
+func (c *clusterImpl) Shrink(ctx context.Context, l *logger.Logger, nodeCount int) error {
+	err := roachprod.Shrink(ctx, l, c.name, nodeCount)
```
True, in my mind shrink should only be useful when everything is "okay" and we want to free up resources. In the context of a `roachtest` it will probably never make sense to call shrink, but rather just stop cockroach etc, and it will likely be at the end of a test. IMHO I'm not sure if there's much value for these in the context of roachtest and we are not entering a space where operations/roachtests don't quite overlap.

herkolategan commented on pkg/roachprod/roachprod.go:
```diff
@@ -1646,7 +1648,24 @@ func Grow(ctx context.Context, l *logger.Logger, clusterName string, numNodes in
+		if err != nil {
+			return err
+		}
+		err = c.RedistributeNodeCert(ctx, l)
```
Good point, we should trigger it here as well. I'll combine the logic a bit as well.

herkolategan replied:
Moved to Cert Distribution.

herkolategan commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2104,6 +2104,26 @@ func (c *clusterImpl) configureClusterSettingOptions(
+	if err != nil {
+		return err
+	}
+	c.spec.NodeCount += nodeCount
```
TBH I haven't thought of this in the context of roachtests. A few MIG related things will probably not fit well in that context yet.

herkolategan replied:
I think I'll keep the Cluster Interface pure (in the sense that it doesn't mutate hardware), and have the operation use `roachprod` directly.

herkolategan commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2104,6 +2104,26 @@ func (c *clusterImpl) configureClusterSettingOptions(
 }
 
+// Grow adds nodes to the cluster.
+func (c *clusterImpl) Grow(ctx context.Context, l *logger.Logger, nodeCount int) error {
```
Not yet, this might be a good future addition, but as you mentioned we would need to address things like quota and possibly other concerns.

herkolategan replied:
Removed this, and changing the resize operation to use roachprod directly.

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -367,21 +367,23 @@ func GrowCluster(l *logger.Logger, c *Cluster, numNodes int) error {
 	providers := c.Clouds()
-	if len(providers) != 1 || providers[0] != gce.ProviderName {
+	provider := c.VMs[0].Provider
+	if len(providers) != 1 || provider != gce.ProviderName {
```
A cluster can be spread across multiple providers, roachprod supports this. Personally I think we should remove the support, as it's not widely used afaik, and generally complicates things.
In the event the first VM's provider here is GCE it doesn't mean all providers of all the VMs will be GCE, the `len(providers)` here is a shorthand to make sure there are no others.

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -367,21 +367,23 @@ func GrowCluster(l *logger.Logger, c *Cluster, numNodes int) error {
 			c.Name, gce.ProviderName)
 	}
 
 	// Only GCE supports expanding a cluster.
```
Done.

herkolategan commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2841,6 +2841,25 @@ func (c *clusterImpl) ExternalIP(
+	}
+	localities := make([]string, 0, len(cachedCluster.VMs))
+	for _, cVM := range cachedCluster.VMs {
+		locality, err := cVM.Locality()
```
I'm removing this, and removing the hacky way to generate an equivalent start script for the new nodes by rather having a config passed with start opts.

herkolategan committed:
roachprod: add managed check to grow command

Epic: None
Release note: None

herkolategan committed:
roachprod: move cert signal to redist certs

Previously, after creating a load balancer we sent a signal to cockroach
processes to reload certs. This functionality should be shared by any call to
`RedistributeNodeCert` since it's likely the cluster will need to reload certs
after new certs are distributed. In the event that cockroach is not running on
the cluster yet, the signal call has no effect, and does not fail.

Similarly, the `grow` operation will need to redistribute all certs to all
nodes, even if certs are already present on node 1. This change also adds the
ability to specify that certs need to be redistributed and a signal sent to all
cockroach processes to reload the certs with all the new node IP info.

Epic: None
Release note: None

herkolategan committed:
roachtest: dynamic cluster interface

The current cluster interface does not allow mutating the cluster topology. This
is correct and it should remain as is. Tests expect an immutable cluster spec
and rely on this not to change, especially since clusters can be re-used.

However, for operations the cluster topology can be mutated to test long-running
clusters in interesting ways. Hence, we introduce a DynamicCluster interface in
this change that add the ability for a cluster to be resized.

Release note: None
Epic: None

herkolategan committed:
roachtest: add operation config

Currently, roachtests have a full view of the cluster spec a test will run on
and since it controls everything that happens on the cluster it is aware of
things like the cluster settings and start options of cockroach.

This is however not true for operations that currently have no way to probe a
cluster for this information. Even through probing it would be hard to derive
things like the start flags of cockroach from the nodes itself.

Hence, we introduce a config that can be passed when running operations. The
config currently covers cluster settings, start options and the cluster spec.

Release note: None
Epic: None

herkolategan committed:
roachtest: add resize operation

This change adds a new resize operation that will grow clusters and start
cockroach on the new nodes. It copies the current binary and libraries from the
first node, and uses the provided operation config to supply start options and
cluster settings when starting cockroach on the new nodes.

In the new state the operation will wait for clean up before draining and
then decommissioning the new nodes.

If the operation succeeded the cleanup function will shrink the cluster back to
its original size. But if the operation fails, the cleanup function is not
invoked and the cluster will remain at its new size. This is useful for
investigating the new nodes for issues, but also should be monitored since if
the operation is run again it will keep increasing the cluster in size.

Release Note: None
Epic: None

herkolategan commented:
@renatolabs Thanks for the review on this, some great points. This PR has changed quite a bit, but the two major changes are isolating the growth and shrink capability to a new cluster interface type `DynamicCluster`, and not using existing start scripts to derive cluster info, but rather add the ability to pass a config to run operations. Should be ready for another pass when you get a chance (TIA).

renatolabs commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -1634,8 +1635,10 @@ func tenantCertsTarName(virtualClusterID int) string {
 // DistributeCerts will generate and distribute certificates to all the nodes.
-func (c *SyncedCluster) DistributeCerts(ctx context.Context, l *logger.Logger) error {
-	if c.checkForCertificates(ctx, l) {
+func (c *SyncedCluster) DistributeCerts(
```
We should add a comment about the new `redistribute` argument. As I understand it, I think it means "did we add any new certs" (implying we want to force redistribution). But spelling it out would be better.

herkolategan replied:
Done.

renatolabs commented on pkg/cmd/roachtest/cluster/dynamic_cluster_interface.go:
```diff
@@ -0,0 +1,25 @@
+
+// DynamicCluster is a cluster that can be resized. This is only for use by
+// operations that mutate the underlying cluster topology.
+type DynamicCluster interface {
```
Do we need an interface for this? Doesn't seem to be used.

herkolategan replied:
This is used in the resize operation, to check if the provided Cluster interface also supports the DynamicCluster interface.

renatolabs reviewed (approved):
LGTM!

Thanks for addressing the comments.

herkolategan commented:
TFTR!

bors r=renatolabs
